internal/fxlog/fxlog.go

```diff
@@ -21,86 +21,127 @@
 package fxlog
 
 import (
+	"strings"
+
+	"go.uber.org/fx/internal/fxreflect"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
-// Level is the level of logging used by Logger.
-type Level int
+// Logger defines interface used for logging.
+type Logger interface {
+	// LogEvent is called when a logging event is emitted.
+	LogEvent(Event)
+}
 
-const (
-	// InfoLevel is Info logging level used to log messages via zap.
-	InfoLevel Level = iota
-	// ErrorLevel is Info logging level used to log messages via zap.
-	ErrorLevel
-)
+// Event defines an event emitted by fx.
+type Event interface{}
 
-// Entry is an entry to be later serialized into zap message and fields.
-type Entry struct {
-	Level   Level
-	Message string
-	Fields  []Field
-	Stack   string
+// LifecycleOnStartEvent is emitted for whenever an OnStart hook is executed
+type LifecycleOnStartEvent struct {
+	Caller string
 }
 
-// WithStack mutates an Entry to add a Stack field.
-func (e Entry) WithStack(stack string) Entry {
-	e.Stack = stack
-	return e
+// LifecycleOnStopEvent is emitted for whenever an OnStart hook is executed
+type LifecycleOnStopEvent struct {
+	Caller string
 }
 
-func (e Entry) Write(logger Logger) {
-	logger.Log(e)
+// ApplyOptionsError is emitted whenever there is an error applying options.
+type ApplyOptionsError struct {
+	Err error
 }
 
-// Err is a helper for error Fields.
-func Err(value error) Field {
-	return F("error", value)
+// SupplyEvent is emitted whenever a Provide was called with a constructor provided
+// by fx.Supply.
+type SupplyEvent struct {
+	Constructor interface{}
```

Comment:
Oh, hold up. fx.Supply doesn't take a function. It takes a value.

```
x := T1{}
fx.Supply(x)
```

We internally turn it into a constructor so that we can call Provide. So it's equivalent to,

```
x := T1{}
fx.Provide(func() T1 { return x })
```

---

For these, we only log the types of values (as seen in the deleted code above). So for now, I would make this a TypeName field.

```suggestion
	TypeName string
```